### ISSUE TITLE
Add restore from backup support

### ISSUE THREAD

**Issue by ants** [zalando/patroni#354]
A feature request I got: when restoring a cluster from backup one has to make sure all Patroni instances are shut down before replacing one data dir from backup, doing the whole recovery dance and possibly adjusting `/optime/leader` and then bringing the nodes back up in correct order. Would be nice to have some tooling for that to reduce room for operator errors and speed things up. Ideally make a restore from backup scriptable.

Something along the lines of:

1. patronictl backup-recovery init - shut down postgresql on all nodes.
2a. patronictl backup-recovery complete - start postgres, make it the leader, update /optime/leader, reinit standby's.
2b. patronictl backup-recovery abort - start rest of the cluster, reinit current node.

I'll leave this as a place holder until the idea has had some time to develop.

**Comment by uspen** [comment]
Hello! Thank for idea!
I use pgBackRest. I restore from backup and run pg_ctl start at first (bc patroni rewriting recovery.conf)

**Comment by s200999900** [comment]
Hello!
I use barman for PITR backups (https://github.com/2ndquadrant-it/barman).

**Comment by carrgilson** [comment]
@s200999900 can you comment on how you've managed to integrate Barman and Patroni?
I'm investigating this and it looks like Patroni will rewrite or override the settings that Barman needs to do its backup?

**Comment by CyberDem0n** [comment]
Restore from backup, PITR and cloning of existing clusters task was floating around for a while and now the time to solve it nicely.

Following user stories are interesting for us:

1. We lost cluster completely and want to restore it from backup until latest wal segment available. Actually it's pretty much working out of the box right now. If the cluster structure is still valid in etcd all nodes will start recovering from backup and one of the nodes will grab the leader lock and promote.

2. We want to do PITR because somebody destroyed data in the database.

3. We want to make a full copy of existing cluster and run it as a separate cluster afterwards.

4. We want to do PITR but to a cluster with another name (Create `new-cluster` from the backup of `existing-cluster`). And after it reached a given time or transaction - promote and run as a separate cluster.

2, 3 and 4 actually possible even now, but requires a lot of manual work, i.e.: creating/removing keys in etcd, stopping and starting nodes, changing/reloading Patroni configuration.
Off course, everything has to be done in the right order and it's easy to make a mistake. 
Also it's not really possible to script it nicely.
The biggest problem here is - after restore is done, cluster must "forget" that it was doing a tricky restore.

Now I will describe the basic idea how can we solve it and what new functionality we need from Patroni. Right now Patroni only looks on `initialize` key in etcd.
If the `initialize` key is there and there is no data directory on the current node, Patroni will bootstrap from the master node (or may be from existing replicas) or from existing backup.

What if we introduce one more key in etcd, named `restore`. This key can contain some information needed for restoring from a backup, like recovery target settings and arguments for a restore script.

How it should work:
1. if there is no data directory and etcd doesn't contain `initialize` key, but there is a `restore` key defined, node must grab `initialize` key and run a restore script instead of running initdb.
1.1 After restoring from backup and having postgres running as a master it probably should change passwords for superuser and replication users.
2. If there are both (`initialize` and `restore`) keys are present - Patroni should wait for the `leader` to get pg_basebackup from it (no changes comparing to current behavior).

Basically `restore` key affects only initial cluster bootstrap process and tells that we should run some external script instead of initdb.

Having such functionality it should would become fairly easy to script all above mentioned cases.

@ants what do you think?

**Comment by CyberDem0n** [comment]
Oh!

Actually it's probably even more simple than that and doesn't even require to have the new key in DCS.

We already have `bootstrap` section in the configuration file, where we can put some initdb arguments, add lines to pg_hba.conf, create users and populate `/config` key in DCS.

So we can just extend this section:
```yaml
bootstrap:
  method: my_bootstrap_method  # if there is no `method` defined - use initdb

  my_bootstrap_method:
    command: /my_bootstrap_command.sh
    restore_command: '/my_restore_command.sh %f %p'
    recovery_target:
      action: promote  # should we enforce it?
#      name:
#      time:
#      xid:
#      inclusive:
#      timeline: latest
```

**Comment by ants** [comment]
Having a way to define custom bootstrap methods looks like a great idea. I
will reflect a bit on if there is anything needed on top of what you
described. But in general it looks like it should cover pretty much
anything the people I have talked with would like to do.

In the longer view I'd like to have a way to register bootstrap
configuration in the DCS and change the config infrastructure so that in
the common case identical config can be shared between nodes (derive name
and connect parameters from hostname). It would simplify initial setup
quite a bit. But that's a separate feature.
